[torch distributed] Implementing reduce_scatter_base

liangluofb opened the pull request:
Summary: Support flattened reduce_scatter.

Test Plan:
buck test mode/opt -c fbcode.enable_gpu_sections=true //caffe2/torch/lib/c10d:ProcessGroupNCCLTest
buck test mode/opt -c fbcode.enable_gpu_sections=true //caffe2/test/distributed:c10d

Differential Revision: D27876281

zhaojuanmao commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -257,7 +258,17 @@ class ProcessGroup : public torch::CustomClassHolder {
       std::vector<at::Tensor>& outputTensors,
       std::vector<std::vector<at::Tensor>>& inputTensors,
-      const ReduceScatterOptions& opts = ReduceScatterOptions()) = 0;
+      const ReduceScatterOptions& = ReduceScatterOptions()) = 0;
```
nit: keep the same as before

Luo-Liang replied:
the linter issued a warning. I will overwrite that.

zhaojuanmao commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -257,7 +258,17 @@ class ProcessGroup : public torch::CustomClassHolder {
+      const ReduceScatterOptions& opts = ReduceScatterOptions())
+      {
+        // does not support it by default
+        throw std::runtime_error("ProcessGroup does not support alltoall");
```
nit: "ProcessGroup does not support _reduce_scatter_base"

Luo-Liang replied:
Thanks!

zhaojuanmao commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -500,6 +500,25 @@ def reduce_scatter(outputs, input_lists, op):
+        pg = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        def reduce_scatter_base(output_t, input_t):
+            work = pg._allgather_base(output_t, input_t)
```
it is testing _allgather_base, not _reduce_scatter_base?

also, would you please add a  test for dist._reduce_scatter_base() API as well? and the error cases

Luo-Liang replied:
sg!

zhaojuanmao replied:
why is '_allgather_base' called here?

zhaojuanmao reviewed (commented):
lgtm overall, some minor comments

zhaojuanmao commented on torch/utils/_cpp_extension_versioner.py:
```diff
@@ -41,13 +41,13 @@ def bump_version_if_changed(self,
-        hash_value = update_hash(hash_value, with_cuda)
-        hash_value = update_hash(hash_value, is_python_module)
-        hash_value = update_hash(hash_value, is_standalone)
+        # hash_value = 0
```
why is this changed?

zhaojuanmao commented on torch/utils/cpp_extension.py:
```diff
@@ -1253,67 +1253,67 @@ def _jit_compile(name,
-
-    if is_standalone:
-        return _get_exec_path(name, build_directory)
+    # if is_python_module and is_standalone:
```
same here, why is this changed?

zhaojuanmao commented:
codes lgtm, will wait until tests are green

zhaojuanmao reviewed (approved):
would you please rebase and make sure tests are green before landing? thanks!

liangluofb committed:
[torch distributed] Implementing reduce_scatter_base (#57567)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/57567

Support flattened reduce_scatter.

Test Plan:
buck test mode/opt -c fbcode.enable_gpu_sections=true //caffe2/torch/lib/c10d:ProcessGroupNCCLTest
buck test mode/opt -c fbcode.enable_gpu_sections=true //caffe2/test/distributed:c10d

Reviewed By: zhaojuanmao

Differential Revision: D27876281

fbshipit-source-id: f63dc2b0484636c2b6af0c3cd64d6806c946fedd
